xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/src/main/webjar/xwiki-link/plugin.js

```diff
@@ -111,6 +111,143 @@
           delete link.attributes['data-freestanding'];
         }
       });
+
+      // Register the the [ autocomplete using the ckeditor mentions plugin.
+      editor.config.mentions = editor.config.mentions || [];
+      editor.config.mentions.push({
+        followingSpace: true,
+        marker: '[',
+        minChars: 0,
+        itemsLimit: 6,
+        feed: function (opts, callback) {
+          // We use the source document to compute the feed URL because we want the suggested link references to be
+          // relative to the edited document (we want the editor to output relative references as much as possible).
+          $.getJSON(editor.config.sourceDocument.getURL('get', $.param({
+            sheet: 'CKEditor.LinkSuggestions',
+            outputSyntax: 'plain',
+            language: XWiki.locale,
+            input: opts.query
+          })), function (data) {
+
+            const uploadItem = {
+              id: "_uploadAttachment",
+              iconClass: "fa fa-upload",
+              iconURL:"",
+              label: editor.localization.get('xwiki-link.upload'),
+              hint: editor.localization.get('xwiki-link.uploadHint')
+            };
+
+
+            // // Upload attachment item should be the first in the list.
+            var callbackData = data;
+            if (uploadItem.label.toLowerCase().startsWith(opts.query.toLowerCase())) {
+              callbackData = [uploadItem].concat(callbackData);
+            }
+
+            callback(callbackData);
+
+          });
+        },
+        itemTemplate: [
+          '<li data-id="{id}" class="ckeditor-autocomplete-item">',
+          '<div>',
+          '<span class="ckeditor-autocomplete-item-icon-wrapper">',
+          // We have to output both icon types but normally only one is defined and the other is hidden.
+          '<img src="{iconURL}"/>',
+          '<span class="{iconClass}"></span>',
+          '</span>',
+          '<span class="ckeditor-autocomplete-item-label">{label}</span>',
+          '</div>',
+          '<div class="ckeditor-autocomplete-item-hint">{hint}</div>',
+          '</li>'
+        ].join(''),
+        outputTemplate: function (item) {
+          // Upload a new attachment.
+          if (item.id === "_uploadAttachment") {
+            require(['jquery'], function ($) {
```

Comment:
You don't need this. You can assume jQuery is loaded already. See the ``var $ = jQuery;`` line at the start.